File: app-desktop/vela-wallet/src/wallet/fixtures.rs
//! Canonical wallet fixtures — the desktop port of
//! `specs/015-wallet-home-ui/data-model.md`. Content is verbatim from the
//! mocks (spec FR-012); chain colors are fixture data, not theme tokens.

use gpui::{Hsla, SharedString, rgb};

use super::{WalletStrings, fill};

pub const WALLET_NAME: &str = "大表哥";
pub const ADDRESS_FULL: &str = "0x14fB1fB21751E29F7Ec48dC450017552E3D1eA5c";
pub const NETWORK_COUNT: u32 = 8;

/// Identicon-board seeds (US3): the cross-platform eyeball parity set.
pub const IDENTICON_BOARD_SEEDS: [&str; 6] = [
    ADDRESS_FULL,
    "0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
    "alice",
    "bob",
    "0x9F3c00000000000000000000000000000000021aE",
    "",
];

pub fn chain_bnb() -> Hsla {
    rgb(0xf0b90b).into()
}
pub fn chain_ethereum() -> Hsla {
    rgb(0x627eea).into()
}
pub fn chain_arbitrum() -> Hsla {
    rgb(0x28a0f0).into()
}
pub fn chain_gnosis() -> Hsla {
    rgb(0x21bca5).into()
}
pub fn chain_base() -> Hsla {
    rgb(0x0052ff).into()
}
pub fn chain_polygon() -> Hsla {
    rgb(0x8247e5).into()
}

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum ActivityKind {
    Sent,
    Received,
    Dapp,
}

#[derive(Clone)]
pub struct ActivityRowModel {
    pub kind: ActivityKind,
    pub title: SharedString,
    pub subtitle: SharedString,
    pub amount: SharedString,
    pub unit: SharedString,
    pub positive: bool,
    pub badge: Hsla,
}

#[derive(Clone)]
pub enum Fiat {
    Value(SharedString),
    NoPrice(SharedString),
    Masked,
}

#[derive(Clone)]
pub struct AssetRowModel {
    pub ticker: SharedString,
    pub chain: SharedString,
    pub badge: Hsla,
    pub balance: SharedString,
    pub fiat: Fiat,
}

#[derive(Clone)]
pub struct ChainRowModel {
    pub name: SharedString,
    /// `None` = the neutral all-networks dot.
    pub dot: Option<Hsla>,
    pub count: u32,
    pub selected: bool,
    /// Which chain this row IS — `None` on the all-networks row, which is the
    /// same `None` the filter itself uses. Carried on the row rather than
    /// derived from its position, because a click has to name a chain and a
    /// position is only a chain until the list re-sorts.
    pub chain_id: Option<u32>,
}

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum BalanceState {
    Normal,
    ZeroLive,
    Loading,
    Hidden,
}

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum StatusKind {
    Warning,
    Refreshing,
}

#[derive(Clone)]
pub struct BalanceModel {
    pub label: SharedString,
    pub state: BalanceState,
    pub integer: SharedString,
    pub decimals: Option<SharedString>,
    pub live: Option<SharedString>,
    pub status: Option<(StatusKind, SharedString)>,
}

pub const MASK: &str = "••••";
pub const BALANCE_MASK: &str = "••••••";

/// The default balance every D-state shows.
pub fn balance_default(s: &WalletStrings) -> BalanceModel {
    BalanceModel {
        label: s.total_balance.clone(),
        state: BalanceState::Normal,
        integer: "$1,383".into(),
        decimals: Some("28".into()),
        live: None,
        status: None,
    }
}

/// Component-board balance variants (gallery Components tab).
pub fn balance_variants(s: &WalletStrings) -> Vec<BalanceModel> {
    vec![
        balance_default(s),
        BalanceModel {
            label: s.total_balance.clone(),
            state: BalanceState::ZeroLive,
            integer: "$0".into(),
            decimals: Some("00".into()),
            live: Some(s.live_indicator.clone()),
            status: None,
        },
        BalanceModel {
            label: s.total_balance.clone(),
            state: BalanceState::Loading,
            integer: "".into(),
            decimals: None,
            live: None,
            status: None,
        },
        BalanceModel {
            label: s.total_balance.clone(),
            state: BalanceState::Hidden,
            integer: BALANCE_MASK.into(),
            decimals: None,
            live: None,
            status: None,
        },
        BalanceModel {
            label: s.total_balance.clone(),
            state: BalanceState::Normal,
            integer: "$1,383".into(),
            decimals: Some("46".into()),
            live: None,
            status: Some((StatusKind::Warning, s.balance_unpriced.clone())),
        },
        BalanceModel {
            label: s.total_balance.clone(),
            state: BalanceState::Normal,
            integer: "$1,383".into(),
            decimals: Some("28".into()),
            live: None,
            status: Some((StatusKind::Refreshing, s.balance_stale.clone())),
        },
    ]
}

/// Which D1 row the celebration is about: the `+120 USDT` receipt at index 1.
///
/// The drawn state and the live one must be about the same thing — a glow on a
/// row the toast is not about would teach the drawing's reader the wrong rule
/// — so the index and the sentence below are two halves of one fixture.
pub const CELEBRATED_ROW: usize = 1;

/// D1b's celebration, in the words the corpus uses for the live one.
///
/// `+120 USDT` is what the row at [`CELEBRATED_ROW`] says, so the pill and the
/// row underneath it agree; formatted here rather than parsed back out of that
/// row's amount string, which is the reverse-parse the core removed.
pub fn receipt_toast(s: &WalletStrings) -> SharedString {
    SharedString::from(fill(
        &fill(&s.toast_received, "amount", "120"),
        "token",
        "USDT",
    ))
}

fn row(
    s: &WalletStrings,
    kind: ActivityKind,
    subtitle: String,
    amount: &str,
    unit: &str,
    positive: bool,
    badge: Hsla,
) -> ActivityRowModel {
    let title = match kind {
        ActivityKind::Sent => s.label_sent.clone(),
        ActivityKind::Received => s.label_received.clone(),
        ActivityKind::Dapp => s.label_dapp.clone(),
    };
    ActivityRowModel {
        kind,
        title,
        subtitle: subtitle.into(),
        amount: amount.into(),
        unit: unit.into(),
        positive,
        badge,
    }
}

/// The four D1 activity rows, timestamps included (desktop subtitles carry
/// `· <day> <clock>` per the D1 mock).
pub fn activity_default(s: &WalletStrings) -> Vec<ActivityRowModel> {
    let today = s.today.as_ref();
    let yesterday = s.yesterday.as_ref();
    vec![
        row(
            s,
            ActivityKind::Sent,
            format!("{} · {today} 14:02", fill(&s.to_name, "name", "hold on")),
            "−2",
            "POL",
            false,
            chain_polygon(),
        ),
        row(
            s,
            ActivityKind::Received,
            format!(
                "{} · {today} 11:20",
                fill(&s.from_name, "name", "0x9F3c…21aE")
            ),
            "+120",
            "USDT",
            true,
            chain_ethereum(),
        ),
        row(
            s,
            ActivityKind::Dapp,
            format!("PancakeSwap · {today} 09:41"),
            "−0.05",
            "BNB",
            false,
            chain_bnb(),
        ),
        row(
            s,
            ActivityKind::Received,
            format!(
                "{} · {yesterday} 20:15",
                fill(&s.from_name, "name", "Alice")
            ),
            "+50",
            "USDC",
            true,
            chain_base(),
        ),
    ]
}

/// Masked variants for the component board (H5's rule: dots, units kept).
pub fn activity_masked(s: &WalletStrings) -> Vec<ActivityRowModel> {
    activity_default(s)
        .into_iter()
        .map(|mut r| {
            r.amount = MASK.into();
            r
        })
        .collect()
}

pub fn assets_default(s: &WalletStrings) -> Vec<AssetRowModel> {
    let value = |v: &str| Fiat::Value(v.into());
    let asset = |ticker: &str, chain: &str, badge: Hsla, balance: &str, fiat: Fiat| AssetRowModel {
        ticker: ticker.into(),
        chain: chain.into(),
        badge,
        balance: balance.into(),
        fiat,
    };
    let _ = s;
    vec![
        asset("BNB", "BNB Chain", chain_bnb(), "0.8533", value("$496.46")),
        asset(
            "ETH",
            "Arbitrum",
            chain_arbitrum(),
            "0.2253",
            value("$422.62"),
        ),
        asset(
            "ETH",
            "Ethereum",
            chain_ethereum(),
            "0.0689",
            value("$129.25"),
        ),
        asset("XDAI", "Gnosis", chain_gnosis(), "74.3965", value("$74.38")),
        asset(
            "USDT",
            "Ethereum",
            chain_ethereum(),
            "53.4836",
            value("$53.48"),
        ),
        asset("USDC", "Polygon", chain_polygon(), "12.04", value("$12.04")),
    ]
}

/// Component-board asset variants: no-price (H4), masked (H5), extremes (H7).
pub fn assets_variants(s: &WalletStrings) -> Vec<AssetRowModel> {
    vec![
        AssetRowModel {
            ticker: "CAKE".into(),
            chain: "BNB Chain".into(),
            badge: chain_bnb(),
            balance: "18.20".into(),
            fiat: Fiat::NoPrice(s.no_price.clone()),
        },
        AssetRowModel {
            ticker: "BNB".into(),
            chain: "BNB Chain".into(),
            badge: chain_bnb(),
            balance: MASK.into(),
            fiat: Fiat::Masked,
        },
        AssetRowModel {
            ticker: "WBTC".into(),
            chain: "以太坊主网 Ethereum".into(),
            badge: chain_ethereum(),
            balance: "0.00000042".into(),
            fiat: Fiat::Value("$0.03".into()),
        },
        AssetRowModel {
            ticker: "USDT".into(),
            chain: "Ethereum".into(),
            badge: chain_ethereum(),
            balance: "1,234,567.8901".into(),
            fiat: Fiat::Value("$1,234,567.89".into()),
        },
    ]
}

pub fn chains(s: &WalletStrings) -> Vec<ChainRowModel> {
    let chain = |name: &str, chain_id: u32, dot: Hsla, count: u32| ChainRowModel {
        name: name.into(),
        dot: Some(dot),
        count,
        selected: false,
        chain_id: Some(chain_id),
    };
    vec![
        ChainRowModel {
            name: s.all_networks.clone(),
            dot: None,
            count: NETWORK_COUNT,
            selected: true,
            chain_id: None,
        },
        chain("BNB Chain", 56, chain_bnb(), 1),
        chain("Ethereum", 1, chain_ethereum(), 3),
        chain("Arbitrum", 42_161, chain_arbitrum(), 1),
        chain("Gnosis", 100, chain_gnosis(), 1),
        chain("Base", 8_453, chain_base(), 1),
        chain("Polygon", 137, chain_polygon(), 1),
    ]
}

/// D3's per-asset activity (BNB): the dApp row plus an older literal-dated one.
pub fn bnb_activity(s: &WalletStrings) -> Vec<ActivityRowModel> {
    let today = s.today.as_ref();
    vec![
        row(
            s,
            ActivityKind::Dapp,
            format!("PancakeSwap · {today} 09:41"),
            "−0.05",
            "BNB",
            false,
            chain_bnb(),
        ),
        row(
            s,
            ActivityKind::Received,
            // 8月1日 is fixture data (a literal date), verbatim per FR-012.
            format!("{} · 8月1日", fill(&s.from_name, "name", "0x21aE…9F3c")),
            "+0.9",
            "BNB",
            true,
            chain_bnb(),
        ),
    ]
}

/// D3, as one model.
///
/// Added in 031 so the live panel and the mock render through one body. The
/// fixture constructor below reproduces the mock's content exactly, so the
/// gallery is unchanged.
#[derive(Clone)]
pub struct AssetDetailModel {
    pub ticker: SharedString,
    pub badge: Hsla,
    /// `0.8533 BNB`.
    pub amount: SharedString,
    /// `$496.46 · BNB Chain`.
    pub sub: SharedString,
    pub facts: Vec<(SharedString, SharedString)>,
    pub activity: Vec<ActivityRowModel>,
}

/// D3 as the mocks draw it.
#[must_use]
pub fn asset_detail_default(s: &WalletStrings) -> AssetDetailModel {
    AssetDetailModel {
        ticker: "BNB".into(),
        badge: chain_bnb(),
        amount: "0.8533 BNB".into(),
        sub: "$496.46 · BNB Chain".into(),
        facts: bnb_facts(s),
        activity: bnb_activity(s),
    }
}

/// D3 fact rows.
pub fn bnb_facts(s: &WalletStrings) -> Vec<(SharedString, SharedString)> {
    vec![
        (s.label_name.clone(), "BNB".into()),
        (
            s.label_price.clone(),
            fill(&fill(&s.price_value, "symbol", "BNB"), "value", "$581.85").into(),
        ),
        (s.label_contract.clone(), s.native_token.clone()),
        (s.label_decimals.clone(), "18".into()),
    ]
}

/// D2 token-picker detail line: `BNB Chain · 链 ID 56`.
pub fn receive_network_detail(s: &WalletStrings) -> SharedString {
    fill(&fill(&s.network_detail, "name", "BNB Chain"), "id", "56").into()
}

/// D2 warning-card footnote: `同一地址，通用于全部 8 个网络`.
pub fn receive_networks_line(s: &WalletStrings) -> SharedString {
    fill(&s.networks_line, "count", &NETWORK_COUNT.to_string()).into()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::loc::Loc;

    /// FR-012: with the zh locale the assembled strings match the mocks
    /// verbatim (this is the desktop twin of the web fixtures test).
    #[test]
    fn zh_fixtures_match_the_mocks() {
        // SAFETY: test-local env pin, same pattern the loc tests rely on.
        unsafe { std::env::set_var("VELA_LANG", "zh") };
        let loc = Loc::from_env();
        let s = WalletStrings::resolve(&loc);

        let rows = activity_default(&s);
        assert_eq!(rows[0].title.as_ref(), "已发送");
        assert_eq!(rows[0].subtitle.as_ref(), "至 hold on · 今天 14:02");
        assert_eq!(rows[3].subtitle.as_ref(), "来自 Alice · 昨天 20:15");

        assert_eq!(receive_network_detail(&s).as_ref(), "BNB Chain · 链 ID 56");
        assert_eq!(
            receive_networks_line(&s).as_ref(),
            "同一地址，通用于全部 8 个网络"
        );

        let facts = bnb_facts(&s);
        assert_eq!(facts[1].1.as_ref(), "1 BNB = $581.85");
        assert_eq!(facts[2].1.as_ref(), "原生代币");

        let chain_rows = chains(&s);
        assert_eq!(chain_rows[0].name.as_ref(), "所有网络");
        assert_eq!(chain_rows.len(), 7);
    }
}

File: app-web/vela-wallet/src/lib/wallet/fixtures.ts
/**
 * Canonical wallet-home fixtures (spec 015, data-model.md — the single canon
 * all four platforms port). Content is verbatim from `docs/design/wallet/` mocks;
 * builders merge it with resolved messages into display-ready view models.
 * Pure data + assembly: no fetching, no formatting rules, no business state.
 */
import { fill, type WalletMessages } from './messages';
import type {
	ActivityGroupModel,
	ActivityRowModel,
	AssetRowModel,
	BalanceModel,
	ChainRowModel,
	DesktopStateId,
	AssetDetailPanelModel,
	MobileStateId,
	NetworkPillModel,
	ReceivePanelModel,
	SectionModel,
	WalletDesktopModel,
	WalletHeaderModel,
	WalletHomeModel
} from './model';

// --- Canon ----------------------------------------------------------------

export const CHAIN_COLORS = {
	bnb: '#F0B90B',
	ethereum: '#627EEA',
	arbitrum: '#28A0F0',
	gnosis: '#21BCA5',
	base: '#0052FF',
	polygon: '#8247E5',
	// Spec 021: the receive list is the first screen to draw all eight
	// supported networks, not just the six the home screen holds balances on.
	optimism: '#FF0420',
	avalanche: '#E84142'
} as const;

export const PILL_DOTS = [CHAIN_COLORS.ethereum, CHAIN_COLORS.polygon, CHAIN_COLORS.bnb];

export const IDENTITY = {
	name: '大表哥',
	longName: '这是一个非常长',
	addressDisplay: '0x14fB1f…D1eA5c',
	addressFull: '0x14fB1fB21751E29F7Ec48dC450017552E3D1eA5c'
} as const;

/** Identicon-board seeds (US3): cross-platform eyeball parity set. */
export const IDENTICON_BOARD_SEEDS = [
	IDENTITY.addressFull,
	'0xd8da6bf26964af9d7eed9e03e53415d37aa96045',
	'alice',
	'bob',
	'0x9F3c00000000000000000000000000000000021aE',
	''
] as const;

/** The privacy masks — exported so the live layer masks with the same glyphs. */
export const MASK = '••••';
export const BALANCE_MASK = '••••••';

/**
 * The live layer's chain colour, by chain id: the canon brand colour for a
 * built-in chain, neutral for anything else. Lives here because chain
 * colours are CONTENT (this file's audit exemption), not tokens.
 */
export function chainColor(chainId: number): string {
	const id = CHAINS_BY_ID[chainId];
	return (
		(id !== undefined ? CHAIN_COLORS[id as keyof typeof CHAIN_COLORS] : undefined) ?? '#8C8C8C'
	);
}

const CHAINS_BY_ID: Record<number, string> = {
	1: 'ethereum',
	56: 'bnb',
	137: 'polygon',
	42161: 'arbitrum',
	8453: 'base',
	100: 'gnosis'
};

type Identicon = (seed: string) => string;

interface ActivityFixture {
	kind: 'sent' | 'received' | 'dapp';
	/** subtitle direction: to/from get localized templates; plain stays verbatim. */
	direction: { to: string } | { from: string } | { plain: string };
	day: 'today' | 'yesterday' | { literal: string };
	clock?: string;
	amount: string;
	unit: string;
	positive: boolean;
	badgeColor: string;
}

const DEFAULT_ACTIVITY: ActivityFixture[] = [
	{
		kind: 'sent',
		direction: { to: 'hold on' },
		day: 'today',
		clock: '14:02',
		amount: '−2',
		unit: 'POL',
		positive: false,
		badgeColor: CHAIN_COLORS.polygon
	},
	{
		kind: 'received',
		direction: { from: '0x9F3c…21aE' },
		day: 'today',
		clock: '11:20',
		amount: '+120',
		unit: 'USDT',
		positive: true,
		badgeColor: CHAIN_COLORS.ethereum
	},
	{
		kind: 'dapp',
		direction: { plain: 'PancakeSwap · BNB Chain' },
		day: 'today',
		clock: '09:41',
		amount: '−0.05',
		unit: 'BNB',
		positive: false,
		badgeColor: CHAIN_COLORS.bnb
	},
	{
		kind: 'received',
		direction: { from: 'Alice' },
		day: 'yesterday',
		clock: '20:15',
		amount: '+50',
		unit: 'USDC',
		positive: true,
		badgeColor: CHAIN_COLORS.base
	}
];

const EXTREME_ACTIVITY: ActivityFixture[] = [
	{
		kind: 'sent',
		direction: { to: 'Alexandra' },
		day: 'today',
		amount: '−1234.5678',
		unit: 'POL',
		positive: false,
		badgeColor: CHAIN_COLORS.polygon
	},
	{
		kind: 'dapp',
		direction: { plain: 'app.uniswap.org · BNB' },
		day: 'today',
		amount: '−0.0000001',
		unit: 'BNB',
		positive: false,
		badgeColor: CHAIN_COLORS.bnb
	}
];

interface AssetFixture {
	ticker: string;
	chain: string;
	badgeColor: string;
	balance: string;
	fiat: string | null;
}

const DEFAULT_ASSETS: AssetFixture[] = [
	{
		ticker: 'BNB',
		chain: 'BNB Chain',
		badgeColor: CHAIN_COLORS.bnb,
		balance: '0.8533',
		fiat: '$496.46'
	},
	{
		ticker: 'ETH',
		chain: 'Arbitrum',
		badgeColor: CHAIN_COLORS.arbitrum,
		balance: '0.2253',
		fiat: '$422.62'
	},
	{
		ticker: 'ETH',
		chain: 'Ethereum',
		badgeColor: CHAIN_COLORS.ethereum,
		balance: '0.0689',
		fiat: '$129.25'
	},
	{
		ticker: 'XDAI',
		chain: 'Gnosis',
		badgeColor: CHAIN_COLORS.gnosis,
		balance: '74.3965',
		fiat: '$74.38'
	},
	{
		ticker: 'USDT',
		chain: 'Ethereum',
		badgeColor: CHAIN_COLORS.ethereum,
		balance: '53.4836',
		fiat: '$53.48'
	},
	{
		ticker: 'USDC',
		chain: 'Polygon',
		badgeColor: CHAIN_COLORS.polygon,
		balance: '12.04',
		fiat: '$12.04'
	}
];

const PARTIAL_PRICE_ASSETS: AssetFixture[] = [
	DEFAULT_ASSETS[0],
	DEFAULT_ASSETS[1],
	{ ticker: 'CAKE', chain: 'BNB Chain', badgeColor: CHAIN_COLORS.bnb, balance: '18.20', fiat: null }
];

const EXTREME_ASSETS: AssetFixture[] = [
	{
		ticker: 'WBTC',
		chain: '以太坊主网 Ethereum',
		badgeColor: CHAIN_COLORS.ethereum,
		balance: '0.00000042',
		fiat: '$0.03'
	},
	{
		ticker: 'USDT',
		chain: 'Ethereum',
		badgeColor: CHAIN_COLORS.ethereum,
		balance: '1,234,567.8901',
		fiat: '$1,234,567.89'
	}
];

interface ChainFixture {
	name: string;
	dot: string;
	count: number;
}

const CHAINS: ChainFixture[] = [
	{ name: 'BNB Chain', dot: CHAIN_COLORS.bnb, count: 1 },
	{ name: 'Ethereum', dot: CHAIN_COLORS.ethereum, count: 3 },
	{ name: 'Arbitrum', dot: CHAIN_COLORS.arbitrum, count: 1 },
	{ name: 'Gnosis', dot: CHAIN_COLORS.gnosis, count: 1 },
	{ name: 'Base', dot: CHAIN_COLORS.base, count: 1 },
	{ name: 'Polygon', dot: CHAIN_COLORS.polygon, count: 1 }
];

export const NETWORK_COUNT = 8;

export const MOBILE_STATES: MobileStateId[] = [
	'h1',
	'h1s',
	'h2',
	'h3',
	'h4',
	'h5',
	'h6',
	'h7',
	'h7x',
	'h8'
];

export const DESKTOP_STATES: DesktopStateId[] = ['d1', 'd2', 'd3'];

// --- Assembly -------------------------------------------------------------

function subtitle(m: WalletMessages, f: ActivityFixture, withTime: boolean): string {
	const direction =
		'to' in f.direction
			? fill(m.activity.toName, { name: f.direction.to })
			: 'from' in f.direction
				? fill(m.activity.fromName, { name: f.direction.from })
				: f.direction.plain;
	if (!withTime) return direction;
	const day =
		f.day === 'today'
			? m.activity.today
			: f.day === 'yesterday'
				? m.activity.yesterday
				: f.day.literal;
	const time = f.clock === undefined ? day : `${day} ${f.clock}`;
	// Desktop dApp rows drop the chain suffix in favor of the timestamp (D1 mock).
	const head = 'plain' in f.direction ? f.direction.plain.split(' · ')[0] : direction;
	return `${head} · ${time}`;
}

function activityRow(
	m: WalletMessages,
	f: ActivityFixture,
	opts: { masked?: boolean; withTime?: boolean } = {}
): ActivityRowModel {
	const title =
		f.kind === 'sent'
			? m.activity.sent
			: f.kind === 'received'
				? m.activity.received
				: m.activity.dapp;
	return {
		kind: f.kind,
		title,
		subtitle: subtitle(m, f, opts.withTime ?? false),
		amount: opts.masked === true ? MASK : f.amount,
		unit: f.unit,
		positive: f.positive,
		masked: opts.masked ?? false,
		badgeColor: f.badgeColor
	};
}

function groupByDay(
	m: WalletMessages,
	fixtures: ActivityFixture[],
	opts: { masked?: boolean; withTime?: boolean } = {}
): ActivityGroupModel[] {
	const groups: ActivityGroupModel[] = [];
	for (const f of fixtures) {
		const label =
			f.day === 'today'
				? m.activity.today
				: f.day === 'yesterday'
					? m.activity.yesterday
					: f.day.literal;
		const last = groups.at(-1);
		const row = activityRow(m, f, opts);
		if (last !== undefined && last.label === label) last.rows.push(row);
		else groups.push({ label, rows: [row] });
	}
	return groups;
}

function assetRow(m: WalletMessages, f: AssetFixture, masked = false): AssetRowModel {
	return {
		ticker: f.ticker,
		chain: f.chain,
		badgeColor: f.badgeColor,
		balance: masked ? MASK : f.balance,
		fiat: masked
			? { kind: 'masked' }
			: f.fiat === null
				? { kind: 'no-price', text: m.balance.noPrice }
				: { kind: 'value', text: f.fiat },
		masked
	};
}

/**
 * The sidebar/sheet network list. Exported so the spec-018 contacts desktop
 * shell feeds the SAME rows into the reused Sidebar without a second copy of
 * the chain canon (SC-006 is about components; this keeps the data single too).
 */
export function chainRowsFor(allNetworksLabel: string): ChainRowModel[] {
	return [
		{ name: allNetworksLabel, dot: 'all', count: NETWORK_COUNT, selected: true },
		...CHAINS.map((c) => ({ name: c.name, dot: c.dot, count: c.count, selected: false }))
	];
}

function chainRows(m: WalletMessages): ChainRowModel[] {
	return chainRowsFor(m.networkFilter.allNetworks);
}

function balance(
	m: WalletMessages,
	state: BalanceModel['state'],
	amount?: { integer: string; decimals: string },
	status?: BalanceModel['status']
): BalanceModel {
	return {
		label: m.balance.totalLabel,
		currency: 'USD',
		state,
		integer: state === 'hidden' ? BALANCE_MASK : amount?.integer,
		decimals: state === 'hidden' ? undefined : amount?.decimals,
		liveText: state === 'zero-live' ? m.balance.liveIndicator : undefined,
		status,
		a11yHide: m.balance.a11yHide,
		a11yShow: m.balance.a11yShow
	};
}

function header(m: WalletMessages, identicon: Identicon, long = false): WalletHeaderModel {
	void m;
	return {
		name: long ? IDENTITY.longName : IDENTITY.name,
		addressDisplay: IDENTITY.addressDisplay,
		addressFull: IDENTITY.addressFull,
		identiconSvg: identicon(IDENTITY.addressFull)
	};
}

const DEFAULT_BALANCE = { integer: '$1,383', decimals: '28' };

/** Assemble the mobile home view model for one H-state. */
export function buildMobileState(
	state: MobileStateId,
	m: WalletMessages,
	identicon: Identicon
): WalletHomeModel {
	const pillAll: NetworkPillModel = {
		kind: 'all',
		dots: [...PILL_DOTS],
		label: m.networkFilter.pillAll
	};
	const sections = (mode: SectionModel['mode']): [SectionModel, SectionModel] => [
		{
			title: m.sections.activity,
			action: m.sections.all,
			mode,
			empty: { title: m.activity.emptyTitle, caption: m.activity.emptyCaption }
		},
		{
			title: m.sections.assets,
			action: m.sections.all,
			mode,
			empty: { title: m.assets.emptyTitle, caption: m.assets.emptyCaption }
		}
	];

	const base: WalletHomeModel = {
		state,
		header: header(m, identicon, state === 'h7' || state === 'h7x'),
		pill:
			state === 'h7' || state === 'h7x'
				? { kind: 'single', dot: CHAIN_COLORS.bnb, label: 'BNB Chain' }
				: pillAll,
		balance: balance(m, 'normal', DEFAULT_BALANCE),
		actions: { receive: m.actions.receive, send: m.actions.send, scan: m.actions.scan },
		activitySection: sections('rows')[0],
		activityGroups: groupByDay(m, DEFAULT_ACTIVITY),
		assetsSection: sections('rows')[1],
		assetRows: DEFAULT_ASSETS.map((f) => assetRow(m, f)),
		tabs: { ...m.nav },
		textScale: state === 'h7x' ? 1.35 : 1
	};

	switch (state) {
		case 'h1':
			// First screen: two activity rows, three asset rows visible (scroll shows the rest).
			return base;
		case 'h1s':
			return base;
		case 'h2': {
			const [activitySection, assetsSection] = sections('empty');
			return {
				...base,
				balance: balance(m, 'zero-live', { integer: '$0', decimals: '00' }),
				activitySection,
				activityGroups: [],
				assetsSection,
				assetRows: []
			};
		}
		case 'h3': {
			const [activitySection, assetsSection] = sections('loading');
			return {
				...base,
				balance: balance(m, 'loading'),
				activitySection,
				activityGroups: [],
				assetsSection,
				assetRows: []
			};
		}
		case 'h4':
			return {
				...base,
				balance: balance(
					m,
					'normal',
					{ integer: '$1,383', decimals: '46' },
					{
						kind: 'warning',
						text: m.balance.unpriced
					}
				),
				activityGroups: groupByDay(m, DEFAULT_ACTIVITY.slice(0, 2)),
				assetRows: PARTIAL_PRICE_ASSETS.map((f) => assetRow(m, f))
			};
		case 'h5':
			return {
				...base,
				balance: balance(m, 'hidden'),
				activityGroups: groupByDay(m, DEFAULT_ACTIVITY, { masked: true }),
				assetRows: DEFAULT_ASSETS.map((f) => assetRow(m, f, true))
			};
		case 'h6':
			return {
				...base,
				balance: balance(m, 'normal', DEFAULT_BALANCE, {
					kind: 'refreshing',
					text: m.balance.stale
				})
			};
		case 'h7':
		case 'h7x':
			return {
				...base,
				balance: balance(m, 'normal', { integer: '$1,234,567', decimals: '89' }),
				activityGroups: groupByDay(m, EXTREME_ACTIVITY),
				assetRows: EXTREME_ASSETS.map((f) => assetRow(m, f))
			};
		case 'h8':
			return { ...base, sheet: { title: m.networkFilter.sheetTitle, rows: chainRows(m) } };
	}
}

/** Assemble the desktop view model for one D-state. */
export function buildDesktopState(
	state: DesktopStateId,
	m: WalletMessages,
	identicon: Identicon
): WalletDesktopModel {
	const receivePanel: ReceivePanelModel = {
		kind: 'receive',
		title: m.receive.title,
		token: {
			ticker: 'BNB',
			badgeColor: CHAIN_COLORS.bnb,
			detail: fill(m.receive.networkDetail, { name: 'BNB Chain', id: 56 })
		},
		qrCaption: m.receive.qrCaption,
		addressLabel: m.receive.addressLabel,
		addressFull: IDENTITY.addressFull,
		copyAddress: m.receive.copyAddress,
		warningTitle: m.receive.warningTitle,
		warningReminder: m.receive.warningReminder,
		networksLine: fill(m.receive.networksLine, { count: NETWORK_COUNT })
	};
	const assetDetailPanel: AssetDetailPanelModel = {
		kind: 'asset-detail',
		title: 'BNB',
		token: {
			ticker: 'BNB',
			badgeColor: CHAIN_COLORS.bnb,
			balance: '0.8533 BNB',
			fiatLine: '$496.46 · BNB Chain'
		},
		send: m.assetDetail.send,
		receive: m.assetDetail.receive,
		facts: [
			{ label: m.assetDetail.labelName, value: 'BNB' },
			{
				label: m.assetDetail.labelPrice,
				value: fill(m.assetDetail.priceValue, { symbol: 'BNB', value: '$581.85' })
			},
			{ label: m.assetDetail.labelContract, value: m.assetDetail.nativeToken },
			{ label: m.assetDetail.labelDecimals, value: '18' }
		],
		viewOnExplorer: m.assetDetail.viewOnExplorer,
		transactionsTitle: m.assetDetail.labelTransactions,
		rows: [
			activityRow(m, DEFAULT_ACTIVITY[2], { withTime: true }),
			activityRow(
				m,
				{
					kind: 'received',
					direction: { from: '0x21aE…9F3c' },
					day: { literal: '8月1日' },
					amount: '+0.9',
					unit: 'BNB',
					positive: true,
					badgeColor: CHAIN_COLORS.bnb
				},
				{ withTime: true }
			)
		]
	};

	return {
		state,
		sidebar: {
			header: header(m, identicon),
			nav: [
				{ id: 'wallet', label: m.nav.wallet, selected: true },
				{ id: 'contacts', label: m.nav.contacts, selected: false },
				{ id: 'explore', label: m.nav.explore, selected: false },
				{ id: 'settings', label: m.nav.settings, selected: false }
			],
			networksTitle: m.sidebar.networks,
			networks: chainRows(m)
		},
		balance: balance(m, 'normal', DEFAULT_BALANCE),
		actions: { receive: m.actions.receive, send: m.actions.send, scan: m.actions.scan },
		activitySection: {
			title: m.sections.activity,
			action: m.sections.all,
			mode: 'rows'
		},
		activityGroups: groupByDay(m, DEFAULT_ACTIVITY, { withTime: true }),
		assetsSection: { title: m.sections.assets, action: m.sections.add, mode: 'rows' },
		assetRows: DEFAULT_ASSETS.map((f) => assetRow(m, f)),
		panels: { receive: receivePanel, assetDetail: assetDetailPanel },
		initialPanel: state === 'd2' ? 'receive' : state === 'd3' ? 'asset-detail' : 'none',
		closeLabel: m.close
	};
}
